c10/core/DispatchKeySet.h

```diff
@@ -120,6 +120,44 @@ class DispatchKeySet final {
 private:
   DispatchKeySet(uint64_t repr) : repr_(repr) {}
   uint64_t repr_ = 0;
+
+public:
+  // STL iterator for DispatchKeySet. Iterates from the first DispatchKey to the
+  // NumDispatchKey enum. The iterate is only invalidated by the destruction of
+  // the underlying DispatchKeySet as the iterator stores a pointer to the raw
+  // represenation of the DispatchKeySet.
+  class iterator {
+   public:
+    using self_type = iterator;
+    using iterator_category = std::input_iterator_tag;
+    using value_type = bool;
+    using difference_type = ptrdiff_t;
+    using pointer = const bool*;
+    using reference = const bool&;
```

Comment:
These seem a bit questionable. What bool are you going to point to? The situation here is analogous to `std::vector`, where `reference` is a proxy class and the pointer is implementation-defined.

Reply:
I wasn't really sure what the `pointer` should be (even if I change the `value_type` to `DispatchKey`). Right now, I'm taking a second pass that only iterates through `DispatchKey`s in the set, but I', still not sure what the pointer should be. My current understanding is that when the iterator is dereferenced a `DispatchKey` value should be returned. However, I'm not sure what a pointer would like in this case; the underlying data representation is a `uint64_t` that we convert to `DispatchKey` when the iterator is dereferenced. Any advice?

Reply:
I'm not sure if you can omit these typedefs, but if you can, that would probably be what I would suggest.